### PR TITLE
[ROCm] Adding support to depthwise_conv_op

### PR DESCRIPTION
**Pull request by jerryyin**
This PR adds ROCm support for the depthwise_conv_ops

The changes in this PR are mostly trivial, please review and merge.

------
Notes:

- CNN for double datatype is not yet supported in ROCm, therefore `dtypes.float64` is ignored
- Test passed: //tensorflow/python/kernel_tests:depthwise_conv_op_test
   - All but testDepthwiseConv2DInputGradFormat and testDepthwiseConv2DFilterGradFormat failing on CPU for `Depthwise convolution on CPU is only supported for NHWC format`

@tatianashp @whchung @chsigg

### CODE REVIEW

File: tensorflow/core/kernels/depthwise_conv_op_gpu.h [hunk 1]
```diff
@@ -192,8 +191,10 @@ __global__ __launch_bounds__(1024, 2) void DepthwiseConv2dGPUKernelNHWCSmall(
   typedef typename detail::PseudoHalfType<T>::Type S;
   assert(CanLaunchDepthwiseConv2dGPUSmall(args));
   // Holds block plus halo and filter data for blockDim.x depths.
-  GPU_DYNAMIC_SHARED_MEM_DECL(8, unsigned char, shared_memory);
   static_assert(sizeof(S) <= 8, "Insufficient alignment detected");
+
```
Comment (chsigg) [inline tensorflow/core/kernels/depthwise_conv_op_gpu.h]: Is this change required?

File: tensorflow/core/kernels/depthwise_conv_op_gpu.h [hunk 2]
```diff
@@ -1123,14 +1127,18 @@ __global__ void __launch_bounds__(640, 2)
 // Device function to compute sub-warp sum reduction for a power-of-two group of
 // neighboring threads.
 template <int kWidth, typename T>
+#if GOOGLE_CUDA
 __device__ __forceinline__ T WarpSumReduce(T val) {
```
Comment (chsigg) [inline tensorflow/core/kernels/depthwise_conv_op_gpu.h]: Is __forceinline__ not supported on ROCm? If it's not, should we introduce a preprocessor macro instead of #if-switching every occurrence.

**Review by chsigg — COMMENTED** [review]
Have you tested this kernel? It is relatively heavy on warp-synchronous programming and might assume 32 lanes per warp.

File: tensorflow/core/kernels/depthwise_conv_op_gpu.h [hunk 3]
```diff
@@ -779,7 +783,7 @@ Status LaunchDepthwiseConv2dGPU(OpKernelContext* ctx, const DepthwiseArgs& args,
   }
 }
 
-// A simple launch pad to launch the Cuda kernel for depthwise convolution.
+// A simple launch pad to launch the Gpu kernel for depthwise convolution.
```
[the remark this answers is not in the document]
Comment (jerryyin) [reply tensorflow/core/kernels/depthwise_conv_op_gpu.h]: Sounds good.

File: tensorflow/core/kernels/depthwise_conv_op_gpu.h (hunk 1 above)
Comment (jerryyin) [reply tensorflow/core/kernels/depthwise_conv_op_gpu.h]: Probably not, will revert

File: tensorflow/core/kernels/depthwise_conv_op_gpu.h (hunk 2 above)
Comment (jerryyin) [reply tensorflow/core/kernels/depthwise_conv_op_gpu.h]: Good catch. I think it is available in ROCm as well, according to [here](https://github.com/ROCm-Developer-Tools/HIP/blob/master/include/hip/hcc_detail/host_defines.h#L55). Will remove the macro.

**Comment by jerryyin** [comment]
In responding to @chsigg's question, 

- Everything run fine except `testDepthwiseConv2DFilterGrad` and `testDepthwiseConv2DFilterGradFormat`. Those two tests are failing due to format unsupported from CPU side. 
- Since ROCm does not satisfy [`#if CUDNN_VERSION >= 7000`](https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/blob/develop-upstream/tensorflow/core/kernels/depthwise_conv_op.cc#L479). `use_cudnn_grouped_conv_ ` is always false. I'm fairly confident that `LaunchDepthwiseConvOp` always runs the local TF kernels so test coverage is good.

It does look like though it is assuming 32 lanes per warp, eg [here](https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/blob/develop-upstream/tensorflow/core/kernels/depthwise_conv_op_gpu.h#L1510). @whchung I don't have enough background to tell what impact this assumption has on AMD Gpus, mind taking a look and remind me what I should do for the default 32 lanes per warp?

**Comment by whchung** [comment]
> It does look like though it is assuming 32 lanes per warp, eg [here](https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/blob/develop-upstream/tensorflow/core/kernels/depthwise_conv_op_gpu.h#L1510). @whchung I don't have enough background to tell what impact this assumption has on AMD Gpus, mind taking a look and remind me what I should do for the default 32 lanes per warp?

I think the reason is because `kCudaWarpAll` is defined as `0xffffffff` so warp voting function would also only activate at most 32 threads on AMD hardware.

**Comment by jerryyin** [comment]
@chsigg I synced with @whchung offline and we agree that the assumption on 32 lanes per warp to be addressed al-together in future PR, together with `kGpuWarpAll` [here](https://github.com/tensorflow/tensorflow/blob/ce09f4ad03a5a97dae2db17b76b95288bbf4bdb1/tensorflow/core/util/gpu_device_functions.h#L137). Testing wise this kernel is doing well in my ROCm machine. Let me know if you have further feedbacks.

File: tensorflow/core/kernels/depthwise_conv_op_gpu.h [hunk 4]
```diff
@@ -323,7 +322,7 @@ __global__ __launch_bounds__(1024, 2) void DepthwiseConv2dGPUKernelNHWCSmall(
   }
 }
 
-// A Cuda kernel to compute the depthwise convolution forward pass
+// A Gpu kernel to compute the depthwise convolution forward pass
```
Comment (chsigg) [inline tensorflow/core/kernels/depthwise_conv_op_gpu.h]: GPU

Comment (jerryyin) [reply tensorflow/core/kernels/depthwise_conv_op_gpu.h]: Ooops, rebased last commit, all Cuda/Gpu updated to GPU in comments

File: tensorflow/core/kernels/depthwise_conv_op.cc [hunk 5]
```diff
@@ -38,10 +38,14 @@ limitations under the License.
 #include "tensorflow/core/util/use_cudnn.h"
 #include "tensorflow/core/util/work_sharder.h"
 
+#if GOOGLE_CUDA || TENSORFLOW_USE_ROCM
```
Comment (dagamayank) [inline tensorflow/core/kernels/depthwise_conv_op.cc]: @jerryyin is `#if GOOGLE_CUDA` below a typo?

Comment (jerryyin) [reply tensorflow/core/kernels/depthwise_conv_op.cc]: @dagamyank Just for clarification, are you talking about line 43 being a typo (redundant)? Line 43 is necessary such that cuda related header is only pulled in under cuda environment. If you are talking about line 41, it is necessary because the include below only pull in GPU related header when necessary. 

Let me know if you mean something else.

**Comment by deven-amd** [comment]
@jerryyin 

can you amend the PR to also enable ROCm support in the `tensorflow/core/kernels/depthwise_conv_grad_op.cc` file. :)

thanks
deven

**Review by whchung — APPROVED** [review]

**Comment by jerryyin** [comment]
A gentle ping @chsigg

**Review by chsigg — APPROVED** [review]

**Review by dagamayank — APPROVED** [review]

**Comment by dagamayank** [comment]
@rthadur can we merge this PR in?
